File: apps/common/telemetry/telemetry.go
package telemetry

import (
	"context"
	"crypto/tls"
	"errors"
	"net/http"

	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/contrib/instrumentation/runtime"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
)

// Config holds the configuration for initializing the tracer.
type Config struct {
	ServiceName    string
	Environment    string
	Endpoint       string
	Token          string
	Dataset        string
	MetricsDataset string
}

// InitTelemetry initializes OpenTelemetry tracing and metrics with Axiom as the backend.
// Returns a shutdown function that should be deferred in main().
func InitTelemetry(ctx context.Context, cfg Config) (func(context.Context) error, error) {
	endpoint := cfg.Endpoint
	if endpoint == "" {
		endpoint = "us-east-1.aws.edge.axiom.co"
	}

	environment := cfg.Environment
	if environment == "" {
		environment = "development"
	}

	metricsDataset := cfg.MetricsDataset
	if metricsDataset == "" {
		metricsDataset = "metrics"
	}

	traceExporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(endpoint),
		otlptracehttp.WithHeaders(map[string]string{
			"Authorization":   "Bearer " + cfg.Token,
			"X-AXIOM-DATASET": cfg.Dataset,
		}),
		otlptracehttp.WithTLSClientConfig(&tls.Config{}),
	)
	if err != nil {
		return nil, err
	}

	metricExporter, err := otlpmetrichttp.New(ctx,
		otlpmetrichttp.WithEndpoint(endpoint),
		otlpmetrichttp.WithHeaders(map[string]string{
			"Authorization":   "Bearer " + cfg.Token,
			"X-AXIOM-DATASET": metricsDataset,
		}),
		otlpmetrichttp.WithTLSClientConfig(&tls.Config{}),
	)
	if err != nil {
		return nil, err
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceNameKey.String(cfg.ServiceName),
			attribute.String("environment", environment),
		),
	)
	if err != nil {
		return nil, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
	)

	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExporter)),
	)

	otel.SetTracerProvider(tp)
	otel.SetMeterProvider(mp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	http.DefaultTransport = otelhttp.NewTransport(http.DefaultTransport)
	if err := runtime.Start(runtime.WithMeterProvider(mp)); err != nil {
		return nil, err
	}

	return func(ctx context.Context) error {
		return errors.Join(tp.Shutdown(ctx), mp.Shutdown(ctx))
	}, nil
}

File: apps/file-service/telemetry.py
# telemetry.py
from opentelemetry import trace, metrics
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.propagate import set_global_textmap
from opentelemetry.propagators.composite import CompositePropagator
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.instrumentation.requests import RequestsInstrumentor


def init_telemetry(
    service_name: str,
    environment: str = "development",
    endpoint: str = "https://us-east-1.aws.edge.axiom.co",
    token: str = None,
    dataset: str = None,
    metrics_dataset: str = "metrics",
):
    if not service_name or not token or not dataset:
        raise ValueError("service_name, token, and dataset are required")

    resource = Resource.create({"service.name": service_name, "environment": environment})


    trace_headers = {
        "Authorization": f"Bearer {token}",
        "X-AXIOM-DATASET": dataset,
    }
    trace_exporter = OTLPSpanExporter(endpoint=endpoint, headers=trace_headers)
    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(tracer_provider)


    metrics_headers = {
        "Authorization": f"Bearer {token}",
        "X-AXIOM-DATASET": metrics_dataset,
    }
    metric_exporter = OTLPMetricExporter(endpoint=endpoint, headers=metrics_headers)
    meter_provider = MeterProvider(
        resource=resource,
        metric_readers=[PeriodicExportingMetricReader(metric_exporter)]
    )
    metrics.set_meter_provider(meter_provider)


    propagator = TraceContextTextMapPropagator()
    set_global_textmap(propagator)

    RequestsInstrumentor().instrument()

    return trace.get_tracer(__name__), meter_provider
